MongoEngine single test failures

There are several mongoengine tests that are not standalone. An example is:

```
(qcf)  daniel  ~/Gits/QCFractal: pytest -vsk test_add_task_queue --tb=short
============================================================================== test session starts ===============================================================================
platform darwin -- Python 3.6.7, pytest-3.8.0, py-1.5.4, pluggy-0.7.1 -- /Users/daniel/anaconda3/envs/qcf/bin/python
cachedir: .pytest_cache
rootdir: /Users/daniel/Gits/QCFractal, inifile: setup.cfg
plugins: cov-2.5.1
collected 102 items / 101 deselected

qcfractal/tests/test_mongoengine_socket.py::test_add_task_queue FAILED

==================================================================================== FAILURES ====================================================================================
______________________________________________________________________________ test_add_task_queue _______________________________________________________________________________
<snip>
E   mongoengine.errors.NotUniqueError: Tried to save duplicate unique keys (E11000 duplicate key error collection: qcf_local_values_test_me.task_queue index: base_result_1 dup key: { : null })
==================================================================== 1 failed, 101 deselected in 2.48 seconds ====================================================================
```

While running the entire mongoengine test suite passes the above. These test should be tagged with a fixture or ensure the test is does not depend on non-local conditions.

--- a/qcfractal/storage_sockets/models.py
+++ b/qcfractal/storage_sockets/models.py
@@ -287,8 +287,8 @@ class TaskQueue(CustomDynamicDocument):
                             choices=['RUNNING', 'WAITING', 'ERROR', 'COMPLETE'])
     manager = db.StringField(default=None)
 
-    created_on = db.DateTimeField(required=True, default=datetime.datetime.now)
-    modified_on = db.DateTimeField(required=True, default=datetime.datetime.now)
+    created_on = db.DateTimeField(required=True, default=datetime.datetime.utcnow)
+    modified_on = db.DateTimeField(required=True, default=datetime.datetime.utcnow)
 
     # can reference Results or any Procedure
     base_result = db.GenericLazyReferenceField(dbref=True)  # use res.id and res.document_type (class)
@@ -316,7 +316,7 @@ class TaskQueue(CustomDynamicDocument):
 
     def save(self, *args, **kwargs):
         """Override save to update modified_on"""
-        self.modified_on = datetime.datetime.now()
+        self.modified_on = datetime.datetime.utcnow()
 
         return super(TaskQueue, self).save(*args, **kwargs)
 
@@ -359,6 +359,7 @@ class QueueManager(CustomDynamicDocument):
     cluster = db.StringField()
     hostname = db.StringField()
     uuid = db.StringField()
+    tag = db.StringField()
 
     # counts
     completed = db.IntField(default=0)
@@ -368,8 +369,8 @@ class QueueManager(CustomDynamicDocument):
 
     status = db.StringField(default='INACTIVE', choices=['ACTIVE', 'INACTIVE'])
 
-    created_on = db.DateTimeField(required=True, default=datetime.datetime.now)
-    modified_on = db.DateTimeField(required=True, default=datetime.datetime.now)
+    created_on = db.DateTimeField(required=True, default=datetime.datetime.utcnow)
+    modified_on = db.DateTimeField(required=True, default=datetime.datetime.utcnow)
 
     meta = {
         'indexes': ['status', 'name', 'modified_on']
@@ -377,6 +378,6 @@ class QueueManager(CustomDynamicDocument):
 
     def save(self, *args, **kwargs):
         """Override save to update modified_on"""
-        self.modified_on = datetime.datetime.now()
+        self.modified_on = datetime.datetime.utcnow()
 
         return super(QueueManager, self).save(*args, **kwargs)

--- a/qcfractal/storage_sockets/models.py
+++ b/qcfractal/storage_sockets/models.py
@@ -325,6 +325,10 @@ class TaskQueue(CustomDynamicDocument):
 
 class ServiceQueue(CustomDynamicDocument):
 
+    status = db.StringField()
+    tag = db.StringField(default=None)
+    hash_index = db.StringField()
+
     meta = {
         'indexes': [
             'status',
@@ -348,13 +352,21 @@ class User(CustomDynamicDocument):
 
 
 class QueueManager(CustomDynamicDocument):
+    """
+    """
 
     name = db.StringField(unique=True)
-    completed = db.IntField()
-    submitted = db.IntField()
-    failures = db.IntField()
-    returned = db.IntField()
-    status = db.StringField(default='INACTIVE')
+    cluster = db.StringField()
+    hostname = db.StringField()
+    uuid = db.StringField()
+
+    # counts
+    completed = db.IntField(default=0)
+    submitted = db.IntField(default=0)
+    failures = db.IntField(default=0)
+    returned = db.IntField(default=0)
+
+    status = db.StringField(default='INACTIVE', choices=['ACTIVE', 'INACTIVE'])
 
     created_on = db.DateTimeField(required=True, default=datetime.datetime.now)
     modified_on = db.DateTimeField(required=True, default=datetime.datetime.now)

--- a/qcfractal/storage_sockets/mongoengine_socket.py
+++ b/qcfractal/storage_sockets/mongoengine_socket.py
@@ -158,44 +158,12 @@ class MongoengineSocket:
         self._tables = self.client[project]
         self._max_limit = max_limit
 
-        # new_table = self.init_database()
-        # for k, v in new_table.items():
-        #     if v:
-        #         self.logger.info("Add '{}' table to the database!".format(k))
+        QueueManager.objects(name='').first()  # init
 
     ### Mongo meta functions
 
     def __str__(self):
         return "<MongoSocket: address='{0:s}:{1:d}:{2:s}'>".format(str(self._url), self._port, str(self._tables_name))
-
-    def init_database(self):
-        """
-        Builds out the initial project structure.
-
-        This is the Mongo definition of "Database"
-        """
-        # Try to create a collection for each entry
-        table_creation = {}
-        # for table in self._valid_tables:
-        #     try:
-        #         # MongoDB "Collection" -> QCFractal "Table"
-        #         self._tables.create_collection(table)
-        #         table_creation[table] = True
-        #
-        #     except pymongo.errors.CollectionInvalid:
-        #         table_creation[table] = False
-
-        # Build the indices
-        # for table, indices in self._table_indices.items():
-        #     idx = [(x, pymongo.ASCENDING) for x in indices if x != "hash_index"]
-        #     self._tables[table].create_index(idx, unique=self._table_unique_indices[table])
-
-        # # Special queue index, hash_index should be unique
-        # for table in ["task_queue", "service_queue"]:
-        #     self._tables[table].create_index([("hash_index", pymongo.ASCENDING)], unique=True)
-
-        # Return the success array
-        return table_creation
 
     def _clear_db(self, db_name: str):
         """Dangerous, make sure you are deleting the right DB"""
@@ -1780,7 +1748,7 @@ class MongoengineSocket:
             if value in kwargs:
                 upd[value] = kwargs[value]
 
-        QueueManager.objects()  # init
+        # QueueManager.objects()  # init
         r = QueueManager.objects(name=name).update(**upd, upsert=True, modified_on=dt.utcnow())
         return r == 1
 

--- a/qcfractal/storage_sockets/mongoengine_socket.py
+++ b/qcfractal/storage_sockets/mongoengine_socket.py
@@ -158,7 +158,6 @@ class MongoengineSocket:
         self._tables = self.client[project]
         self._max_limit = max_limit
 
-        QueueManager.objects(name='').first()  # init
 
     ### Mongo meta functions
 
@@ -168,7 +167,7 @@ class MongoengineSocket:
     def _clear_db(self, db_name: str):
         """Dangerous, make sure you are deleting the right DB"""
 
-        logging.warning('Clearing the Databse, droping tables and the whole DB.')
+        logging.warning('Clearing the Databse and droping tables')
 
         # make sure it's the right DB
         if get_db().name == db_name:
@@ -183,6 +182,8 @@ class MongoengineSocket:
             QueueManager.drop_collection()
 
             self.client.drop_database(db_name)
+
+        QueueManager.objects(name='').first()  # init
 
     def get_project_name(self):
         return self._project_name

--- a/qcfractal/storage_sockets/mongoengine_socket.py
+++ b/qcfractal/storage_sockets/mongoengine_socket.py
@@ -1363,14 +1363,18 @@ class MongoengineSocket:
         """
 
         update = {}
+        not_allowed_keys = []
         # create safe query with allowed keys only
         # shouldn't be allowed to update status manually
         for key, value in data.items():
             if key not in ['procedure', 'program', 'status', 'task_id']:  # FIXME: what else?
                 update[key] = value
             else:
-                logging.warning('Trying to update Procedre with none ' +
-                                'allowed keyword: ({}={})'.format(key, value))
+                not_allowed_keys.append(key)
+
+        if not_allowed_keys:
+            logging.warning('Trying to update Procedre with none ' +
+                            'allowed keyword: {}'.format(not_allowed_keys))
 
         modified_count = Procedure.objects(hash_index=hash_index).update(**update, modified_on=dt.utcnow())
 

--- a/qcfractal/storage_sockets/mongoengine_socket.py
+++ b/qcfractal/storage_sockets/mongoengine_socket.py
@@ -168,6 +168,8 @@ class MongoengineSocket:
     def _clear_db(self, db_name: str):
         """Dangerous, make sure you are deleting the right DB"""
 
+        logging.warning('Clearing the Databse, droping tables and the whole DB.')
+
         # make sure it's the right DB
         if get_db().name == db_name:
             logging.info('Clearing database: {}'.format(db_name))
@@ -1735,7 +1737,7 @@ class MongoengineSocket:
 
     def manager_update(self, name, **kwargs):
 
-        upd = {
+        inc_count = {
             # Increment relevant data
             "inc__submitted": kwargs.pop("submitted", 0),
             "inc__completed": kwargs.pop("completed", 0),
@@ -1743,14 +1745,18 @@ class MongoengineSocket:
             "inc__failures": kwargs.pop("failures", 0)
         }
 
-        # Update server data
-        for value in ["cluster", "hostname", "uuid", "tag", "status"]:
-            if value in kwargs:
-                upd[value] = kwargs[value]
+        upd = {key: kwargs[key] for key in QueueManager._fields_ordered if key in kwargs}
 
         # QueueManager.objects()  # init
-        r = QueueManager.objects(name=name).update(**upd, upsert=True, modified_on=dt.utcnow())
-        return r == 1
+        manager = QueueManager.objects(name=name)
+        if manager:  # existing
+            upd.update(inc_count)
+            num_updated = manager.update(**upd, modified_on=dt.utcnow())
+        else:  # create new, ensures defaults and validations
+            QueueManager(name=name, **upd).save()
+            num_updated = 1
+
+        return num_updated == 1
 
     def get_managers(self, name: str=None, status: str=None, modified_before=None):
 

--- a/qcfractal/storage_sockets/mongoengine_socket.py
+++ b/qcfractal/storage_sockets/mongoengine_socket.py
@@ -1341,8 +1341,8 @@ class MongoengineSocket:
                 not_allowed_keys.append(key)
 
         if not_allowed_keys:
-            logging.warning('Trying to update Procedre with none ' +
-                            'allowed keyword: {}'.format(not_allowed_keys))
+            logging.warning('Trying to update Procedure immutable keywords ' +
+                            '"{}", skipping'.format(not_allowed_keys))
 
         modified_count = Procedure.objects(hash_index=hash_index).update(**update, modified_on=dt.utcnow())
 

--- a/qcfractal/testing.py
+++ b/qcfractal/testing.py
@@ -295,7 +295,6 @@ def reset_server_database(server):
     """Resets the server database for testing.
     """
     server.storage._clear_db(server.storage._project_name)
-    # server.storage.init_database()
 
 
 @pytest.fixture(scope="module")

--- a/qcfractal/tests/test_authentication.py
+++ b/qcfractal/tests/test_authentication.py
@@ -44,7 +44,6 @@ def sec_server(request):
 
         # Clean and re-init the databse
         server.storage.client.drop_database(server.storage._project_name)
-        server.storage.init_database()
 
         # Add local users
         for k, v in _users.items():

--- a/qcfractal/tests/test_mongoengine_socket.py
+++ b/qcfractal/tests/test_mongoengine_socket.py
@@ -11,6 +11,22 @@ from qcfractal.storage_sockets.models import Molecule, Result, Options, \
                     Procedure, OptimizationProcedure, TorsiondriveProcedure
 from qcfractal.storage_sockets.models import TaskQueue
 from qcfractal.testing import mongoengine_socket_fixture as storage_socket
+import pytest
+from bson import ObjectId
+
+
+@pytest.fixture
+def molecules_H4O2(storage_socket):
+    water = portal.data.get_molecule("water_dimer_minima.psimol")
+    water2 = portal.data.get_molecule("water_dimer_stretch.psimol")
+
+    ret = storage_socket.add_molecules({"water1": water.json(as_dict=True),
+                                        "water2": water2.json(as_dict=True)})
+
+    yield list(ret['data'].values())
+
+    storage_socket.del_molecules(water.get_hash(), index="hash")
+    storage_socket.del_molecules(water2.get_hash(), index="hash")
 
 
 def test_molecule(storage_socket):
@@ -57,8 +73,12 @@ def test_molecule(storage_socket):
                                molecular_formula=water_mol.molecular_formula)
     assert len(one_mol) == 1
 
+    # Clean up
+    storage_socket.del_molecules(water.get_hash(), index="hash")
+    storage_socket.del_molecules(water2.get_hash(), index="hash")
 
-def test_results(storage_socket):
+
+def test_results(storage_socket, molecules_H4O2):
     """
         Handling results throught the ME classes
     """
@@ -66,12 +86,12 @@ def test_results(storage_socket):
     assert Result.objects().count() == 0
     assert Options.objects().count() == 0
 
-    molecules = Molecule.objects(molecular_formula='H4O2')
+    molecules = molecules_H4O2
 
-    assert molecules.count() == 2
+    assert len(molecules) == 2
 
     page1 = {
-        "molecule": molecules[0],
+        "molecule": ObjectId(molecules[0]),
         "method": "M1",
         "basis": "B1",
         "options": None,
@@ -81,7 +101,7 @@ def test_results(storage_socket):
     }
 
     page2 = {
-        "molecule": molecules[1],
+        "molecule": ObjectId(molecules[1]),
         "method": "M1",
         "basis": "B1",
         "options": None,
@@ -90,10 +110,12 @@ def test_results(storage_socket):
         "other_data": 10,
     }
 
-    result = Result(**page1)
-    result.save()
-    # print('Result After save: ', result.json(as_dict=True))
-    assert result.molecule.molecular_formula == 'H4O2'
+    Result(**page1).save()
+    ret = Result.objects(method='M1').first()
+    assert ret.molecule.molecular_formula == 'H4O2'
+
+    # clean up
+    Result.objects().delete()
 
 
 def test_procedure(storage_socket):
@@ -122,11 +144,12 @@ def test_procedure(storage_socket):
 
     procedure = Procedure(**data1)
     procedure.save()
+    assert procedure.id
     # print('Procedure After save: ', procedure.json(as_dict=True))
     # assert procedure.molecule.molecular_formula == 'H4O2'
 
 
-def test_optimization_procedure(storage_socket):
+def test_optimization_procedure(storage_socket, molecules_H4O2):
     """
         Optimization procedure
     """
@@ -134,10 +157,8 @@ def test_optimization_procedure(storage_socket):
     assert OptimizationProcedure.objects().count() == 0
     # assert Options.objects().count() == 0
 
-    molecules = Molecule.objects(molecular_formula='H4O2')
-
     data1 = {
-        "initial_molecule": molecules[0],
+        "initial_molecule": ObjectId(molecules_H4O2[0]),
         # "procedure_type": None,
         "options": None,
         "program": "P7",
@@ -149,10 +170,9 @@ def test_optimization_procedure(storage_socket):
         },
     }
 
-    procedure = OptimizationProcedure(**data1)
-    procedure.save()
-    # print('OptimizationProcedure After save: ', procedure.json(as_dict=True))
-    assert procedure.initial_molecule.molecular_formula == 'H4O2'
+    procedure = OptimizationProcedure(**data1).save()
+    proc = OptimizationProcedure.objects().first()
+    assert proc.initial_molecule.molecular_formula == 'H4O2'
 
 
 def test_torsiondrive_procedure(storage_socket):
@@ -184,7 +204,7 @@ def test_torsiondrive_procedure(storage_socket):
     # print('TorsiondriveProcedure After save: ', procedure.json(as_dict=True))
 
 
-def test_add_task_queue(storage_socket):
+def test_add_task_queue(storage_socket, molecules_H4O2):
     """
         Simple test of adding a task using the ME classes
         in QCFractal, tasks should be added using storage_socket
@@ -193,8 +213,17 @@ def test_add_task_queue(storage_socket):
     assert TaskQueue.objects.count() == 0
     TaskQueue.objects().delete()
 
+    page1 = {
+        "molecule": ObjectId(molecules_H4O2[0]),
+        "method": "M1",
+        "basis": "B1",
+        "options": None,
+        "program": "P1",
+        "driver": "energy",
+        "other_data": 5,
+    }
     # add a task that reference results
-    result = Result.objects().first()
+    result = Result(**page1).save()
 
     task = TaskQueue(base_result=result)
     task.save()
@@ -208,7 +237,18 @@ def test_add_task_queue(storage_socket):
     assert TaskQueue.objects().count() == 2
 
     # add a task that reference Torsiondrive Procedure
-    tor = TorsiondriveProcedure.objects().first()
+    data1 = {
+        "options": None,
+        "program": "P9",
+        "qc_meta": {
+            "basis": "B1",
+            "program": "P1",
+            "method": "M1",
+            "driver": "energy"
+        },
+    }
+
+    tor = TorsiondriveProcedure(**data1).save()
 
     task = TaskQueue(base_result=tor)
     task.save()

